specs/player_spec.rb

```diff
@@ -0,0 +1,155 @@
+      @person.play("")
+      @person.play(first_word)
+      @person.play(555)
+      @person.total_score.must_equal Scrabble::Scoring.score(first_word)
+    end
+  end
+
+  describe "won? method" do
+    it "Recognizes when someone wins the Scrabble game" do
+      2.times do
+        hand = @person.draw_tiles(@tilebag)
+        @person.play(hand.join)
+      end
+      @person.won?.must_equal true
+    end
+
+    it "Recognizes when someone has not won the Scrabble game" do
+      first_hand = @person.draw_tiles(@tilebag)
+      @person.play(first_hand.sample(rand(1..7)).join)
+      @person.won?.must_equal false
+    end
+  end
+
+  describe "highest_scoring_word method" do
+    it "Gives the word with the highest score" do
+      first_hand = @person.draw_tiles(@tilebag)
+      first_word = first_hand.join
+      @person.play(first_word)
+      second_hand = @person.draw_tiles(@tilebag)
+      @person.play(second_hand.sample(3).join)
+      third_hand = @person.draw_tiles(@tilebag)
+      @person.play(third_hand.sample(4).join)
+      @person.highest_scoring_word.must_equal first_word
+    end
+  end
+
+  describe "highest_word_score method" do
+    it "Gives the highest score received for a word" do
+      first_hand = @person.draw_tiles(@tilebag)
+      first_word = first_hand.join
+      @person.play(first_word)
+      second_hand = @person.draw_tiles(@tilebag)
+      @person.play(second_hand.sample(3).join)
+      third_hand = @person.draw_tiles(@tilebag)
+      @person.play(third_hand.sample(4).join)
+      @person.highest_word_score.must_equal Scrabble::Scoring.score(first_word)
+    end
+  end
+
+  describe "tiles variable" do
+    it "Is an Array" do
+      @person.tiles.must_be_kind_of Array
+    end
+
+    it "Does not have more than 7 tiles" do
+      @person.tiles.length.must_be :<=, 7
+    end
+  end
+
+  describe "draw_tiles(tilebag) method" do
+    it "Fills empty tiles Array with 7 letters" do
+      @person.draw_tiles(@tilebag)
+      @person.tiles.length.must_equal 7
+    end
+
+    it "Fills tiles Array after playing a word with 7 letters" do
+      @person.draw_tiles(@tilebag)
+      @person.play("#{@person.tiles[0]}#{@person.tiles[1]}")
+      @person.draw_tiles(@tilebag)
+      @person.tiles.length.must_equal 7
+    end
+
+    it "Does not give tiles when hand is full" do
+      first_hand = @person.draw_tiles(@tilebag)
+      second_hand = @person.draw_tiles(@tilebag)
+      first_hand.must_equal second_hand
+    end
+  end
+
+# Optional Enhancements tests
+  describe "valid? method" do
+    it "Will return false when player plays letters not in their hand" do
+      alphabet = @tilebag.tiles.keys.map { |letter| letter.to_s }
+      user_hand = @person.draw_tiles(@tilebag)
+      not_hand = []
+      rand(1..7).times do
+        letter = alphabet.sample(1)
+        not_hand << letter unless user_hand.include? letter
+      end
+      @person.valid?((not_hand).join).must_equal false
```

Comment:
I think this test has a random chance of failing due to the `times` loop above. Here's an example scenario where it might not work:

1. `rand(1..7)` returns 1.
1. `alphabet.sample(1)` returns "A".
1. `user_hand.include? letter` returns true because `@person.tiles` happens to include that letter.
1. We don't push anything into `not_hand` so when we call `valid?(not_hand.join)` we're actually calling it on an empty string.
1. `valid?` as currently written will always return true for an empty string, so this test fails.

What we need to do is find a more consistent way to select a word which contains at least one letter that is not in the player's tile array. One approach to that could be to take `alphabet` and remove all of the letters in `user_hand` from it, leaving us only the letters that were not in `user_hand`.

Also, we should probably have an additional test for `valid?` which verifies that an empty string is not a valid word.